include/seqan3/core/metafunction/range.hpp

```diff
@@ -57,6 +57,16 @@
 //NOTE(h-2): for the range overloads we explicitly forbid that the type is iteratoer
 // because some types are actually both (e.g. std::directory_iterator)
 
+namespace seqan3::detail
+{
+
+//!\cond
+template <typename t>
+concept has_value_type = requires { typename value_type_t<remove_cvref_t<t>>; };
```

Comment:
1) Why the extra concept? I remember from one of the meeting C++ talks that `requires requires` clauses are not recommended but is this the reason or was there a specialization issue?

2) If we do have an extra concept shouldn't this go into `/core` and be used in all similar places?

Reply:
1.
    Here, the problem is that we use the overload-resolution property of concepts. 

    If we don't use a single concept `detail::has_value_type` here, the compiler does not now that both statements are the same (the definition of "same" is if the same line number in the same file of a concept matches) and `innermost_value_type` would not see that the second definition is a more specialized version of the first one.

    ```c++
    template <typename t>
        requires detail::has_value_type<t>
    struct innermost_value_type

    template <typename t>
        requires detail::has_value_type<t> && detail::has_value_type<value_type_t<remove_cvref_t<t>>>
    struct innermost_value_type<t>
    ```

2. The extra concept is only local, therefore in the same file. If we need that in other file we can think about the location.